src/main/java/net/fabricmc/loom/task/service/UnpickService.java

```diff
@@ -108,73 +132,101 @@ public UnpickService(Options options, ServiceFactory serviceFactory) {
-			params.getMainClass().set("daomephsta.unpick.cli.Main");
-			params.getArgs().set(args);
-		});
-
-		workQueue.await();
-
-		return outputJar;
-	}
-
-	private List<String> getUnpickArgs(Path inputJar, Path outputJar, @Nullable Path existingClasses) {
-		var fileArgs = new ArrayList<File>();
-
-		fileArgs.add(inputJar.toFile());
-		fileArgs.add(outputJar.toFile());
-		fileArgs.add(getOptions().getUnpickDefinitions().get().getAsFile());
-		fileArgs.add(getOptions().getUnpickConstantJar().getSingleFile());
-
-		for (File file : getOptions().getUnpickClasspath()) {
-			fileArgs.add(file);
-		}
 
-		if (existingClasses != null) {
-			fileArgs.add(existingClasses.toFile());
+		Files.deleteIfExists(outputJar);
+
+		try (ZipFsClassResolver classResolver = ZipFsClassResolver.create(classpath);
+				InputStream unpickDefinitions = Files.newInputStream(unpickDefinitionsPath)) {
+			ConstantUninliner uninliner = ConstantUninliner.builder()
+					.logger(JAVA_LOGGER)
+					.classResolver(classResolver)
+					.grouper(ConstantGroupers.dataDriven()
+							.logger(JAVA_LOGGER)
+							.classResolver(classResolver)
+							.mappingSource(unpickDefinitions)
+							.build())
+					.build();
+
+			AsyncZipProcessor.processEntries(inputJar, outputJar, new UnpickZipProcessor(uninliner));
 		}
 
-		return fileArgs.stream()
-				.map(File::getAbsolutePath)
-				.toList();
+		return outputJar;
 	}
 
 	public String getUnpickCacheKey() {
 		return Checksum.of(List.of(
 				Checksum.of(getOptions().getUnpickDefinitions().getAsFile().get()),
-				Checksum.of(getOptions().getUnpickConstantJar()),
-				Checksum.of(getOptions().getUnpickRuntimeClasspath())
+				Checksum.of(getOptions().getUnpickConstantJar())
 		)).sha256().hex();
 	}
 
-	public interface UnpickParams extends WorkParameters {
-		Property<String> getMainClass();
-		ListProperty<String> getArgs();
+	private record UnpickZipProcessor(ConstantUninliner uninliner) implements AsyncZipProcessor {
+		@Override
+		public void processEntryAsync(Path input, Path output) throws IOException {
+			Files.createDirectories(output.getParent());
+
+			String fileName = input.toAbsolutePath().toString();
+
+			if (!fileName.endsWith(".class")) {
+				// Copy non-class files
+				Files.copy(input, output);
+				return;
+			}
+
+			ClassNode classNode = new ClassNode();
+
+			try (InputStream is = Files.newInputStream(input)) {
+				ClassReader reader = new ClassReader(is);
+				reader.accept(classNode, 0);
+			}
+
+			LOGGER.debug("Unpick class: {}", classNode.name);
+			uninliner.transform(classNode);
+
+			ClassWriter writer = new ClassWriter(ClassWriter.COMPUTE_MAXS);
+			classNode.accept(writer);
+
+			Files.write(output, writer.toByteArray());
+		}
 	}
 
-	public abstract static class UnpickAction implements WorkAction<UnpickParams> {
-		private static final Logger LOGGER = LoggerFactory.getLogger(UnpickAction.class);
+	private static class ZipFsClassResolver extends ChainClassResolver implements Closeable {
```

Comment:
`ChainClassResolver` is an impl class, I think it's better to have this class implement `IClassResolver` directly and handle one zip file system, then `create` can return an `IClassResolver` by using `IClassResolver.chain`.

Reply:
Good spot, now fixed. Thanks.